components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java

```diff
@@ -2042,9 +2042,30 @@ public int compare(AuthenticationStep step1, AuthenticationStep step2) {
 
             Arrays.sort(authenticationSteps, comparator);
 
+            int numSteps = authenticationSteps.length;
+            // We check if the steps have consecutive step numbers
+            if (numSteps > 0 && authenticationSteps[numSteps-1].getStepOrder() != numSteps) {
```

Comment:
Can you make sure steps are in order by just looking at the last step ? . Can't it go wrong in between other steps ?

Reply:
We do a sort of the steps in line 2043. So the steps will in order according to their step order. So the logic here is not to check whether the steps are in order but rather to identify if the steps having any gaps in between.

Lets say we have steps 1,2,3,4 after force delete step 2 is removed. Now we have 1,3,4. After sorting we would have 1,3,4. Logic here is we look at the step order of the last step which is 4 and compare it with the number of steps which is 3. So this tells us there is a gap in between. 

So we iterate and assign step number so that 1,3,4 --> 1,2,3

Reply:
Add logs here